subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -375,9 +375,21 @@ DefaultJavaCompileSpec createSpec() {
 
     private void configureCompatibilityOptions(DefaultJavaCompileSpec spec) {
         if (javaCompiler.isPresent()) {
-            final JavaInstallationMetadata toolchain = javaCompiler.get().getMetadata();
-            spec.setTargetCompatibility(toolchain.getLanguageVersion().asString());
-            spec.setSourceCompatibility(toolchain.getLanguageVersion().asString());
+            if (compileOptions.getRelease().isPresent()) {
+                spec.setRelease(compileOptions.getRelease().get());
+            } else {
+                final JavaInstallationMetadata toolchain = javaCompiler.get().getMetadata();
+                if (super.getSourceCompatibility() != null) {
+                    spec.setSourceCompatibility(getSourceCompatibility());
+                } else {
+                    spec.setSourceCompatibility(toolchain.getLanguageVersion().toString());
```

Comment:
I find it scary that we rely on toString instead of a method that properly declares what it returns

Reply:
Interesting, I had `asString()` in there and changed it back.
In this case, `toString` is documented and explain what it returns though 😉